pkg/upgrade/key_pair.go

```diff
@@ -8,42 +8,55 @@ import (
 	"strings"
 
 	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
-	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/certificates"
-	clusterapiv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
 )
 
-// KeyPair contains a cert and key.
-type KeyPair struct {
+// KubeconfigSecretKey is the key where the kubeconfig is stored in the secret.
+const KubeconfigSecretKey = "kubeconfig"
+
+// keyPair contains a cert and key.
+type keyPair struct {
 	Cert []byte `json:"cert"`
 	Key  []byte `json:"key"`
 }
 
-type keyPairGetter struct {
-	secretClient corev1client.SecretsGetter
+// secrets has already been scoped to namespace, so, .Secrets("my-namespace") returns one of these.
+type secrets interface {
+	Get(string, metav1.GetOptions) (*v1.Secret, error)
 }
 
-func newKeyPairGetter(secretClient corev1client.SecretsGetter) *keyPairGetter {
-	return &keyPairGetter{
-		secretClient: secretClient,
+// NewRestConfigFromKubeconfigSecretRef decodes the kubeconfig stored in a secret and builds a *rest.Config with it.
+func NewRestConfigFromKubeconfigSecretRef(secrets secrets, name string) (*rest.Config, error) {
+	secret, err := secrets.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
 	}
-}
 
-func (k *keyPairGetter) getKeyPair(cluster *clusterapiv1alpha1.Cluster, config KeyPairConfig) (*KeyPair, error) {
-	if config.ClusterField != "" {
-		return getEmbeddedKeyPair(cluster, config.ClusterField)
+	kc, ok := secret.Data[KubeconfigSecretKey]
+	if !ok {
+		return nil, errors.Errorf("item 'kubeconfig' not found in secret %q", name)
 	}
-
-	return k.getSecretRefKeyPair(cluster.Namespace, config.SecretRef)
+	out := make([]byte, base64.StdEncoding.DecodedLen(len(kc)))
+	n, err := base64.StdEncoding.Decode(out, kc)
```

Comment:
this actually doesn't need to be decoded, the client does that automatically which really confused the heck out of me.

Reply:
Does this mean you need to remove this `Decode` call?

Reply:
yes, that is correct

Reply:
Actual result:

* Do not store secrets that you encode with base64. `kubectl` automatically encodes the data. If you *do* store b64 secrets, you will need to read them with `kubectl get secret ... | base64 --decode | base64 --decode`.
* Assume that providers are storing secrets that are *not* b64 encoded.